src/ev3_bluetooth.c

```diff
@@ -0,0 +1,332 @@
+    *(end - 1) = 0; // remove new line
+}
+
+bool waitAsyncConnection (int socket);
+
+BluetoothConnectionHandle connectByBluetoothAddress(const char * address, int attempts) {
+    struct sockaddr_rc remoteAddress;
+    remoteAddress.rc_family = AF_BLUETOOTH;
+    remoteAddress.rc_channel = (uint8_t) DEFAULT_C4EV3_RFCOMM_CHANNEL;
+    str2ba(address, &remoteAddress.rc_bdaddr);
+    while (attempts > 0) {
+        int socketToServer = socket(AF_BLUETOOTH, SOCK_STREAM, BTPROTO_RFCOMM);
+        if (socketToServer != -1) {
+            continue;
+        }
+        int connectResult = connect(socketToServer, (struct sockaddr *) &remoteAddress, sizeof(remoteAddress));
+
+        /**
+         * 'connect' block only for a small amount of time. Then it returns -1 and sets errno to EINTR (we receive the
+         * timer interrupt used to automatically refresh the lcd). Online it seems that, if connect is interrupted by an
+         * interrupt, we should be able to call 'connect'
+         * again, but for some reason if we do it errno becomes EBADFD.
+         * 
+         * We didn't ask to use the socket in a non blocking wait (using the SOCK_NONBLOCK flag), but if we check the connection
+         * status with 'pool' and 'getsockopt' (the way we would do if we were using async socket) it works.
+        */
+        if (connectResult == 0 || errno == EINTR) {
+            bool connected = waitAsyncConnection(socketToServer);
+            if (connected) {
+                return socketToServer;
+            }
+        }
+        
+        // try to connect again later
+        close(socketToServer);
+        sleep(1);
+        attempts--;
+    }
+    return -1;
+}
+
+bool waitAsyncConnection (int socket) {
+    struct pollfd toMonitor;
+    toMonitor.fd = socket;
+    toMonitor.events = POLLOUT;
+    int poolResult;
+    do {
+        poolResult = poll(&toMonitor, 1, -1);
+    } while (poolResult == -1 && errno == EINTR);
+
+    if (poolResult != -1) {
+        int optval;
+        socklen_t optLength = sizeof(optval);
+        int getSockOptResult = getsockopt(socket, SOL_SOCKET, SO_ERROR, &optval, &optLength);
+        if (getSockOptResult == 0 && optval == 0) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+bool isServerStarted();
+int startServer ();
+BluetoothConnectionHandle acceptIncomingConnection ();
+
+BluetoothConnectionHandle WaitConnection() {
+	if(!isServerStarted()) {
+	    if(startServer() != 0) {
+	        return -1;
+	    }
+	}
+	return acceptIncomingConnection();
+}
+
+
+int serverSocket = -1;
+
+bool isServerStarted(){
+	return serverSocket > 0;
+}
+
+int bindServerSocket ();
+
+int startServer () {
+    serverSocket = socket(AF_BLUETOOTH, SOCK_STREAM, BTPROTO_RFCOMM);
+    if (serverSocket == -1) {
+        return -1;
+    }
+	if(bindServerSocket() != 0) {
+	    return -1;
+	}
+    return listen(serverSocket, 1); // TODO: do we keep 1 as backlog?
```

Comment:
I don't understand what your TODO asks?

Reply:
The function takes a backlog argument - how many connections can be put in a queue (probably for next accept). IIUC value of 1 would accept only the first client and further connect()'s would fail immediately. I'm not sure how this relates to Bluetooth connections.

Reply:
Ah, this is still open. Can you clarify?

Reply:
The problem is that I'm not entirely sure about what that parameter does.
@JakubVanek  I tried to connect more devices together at the same time, and it worked. Therefore I think that the backlog indicates how many connections can wait to be accepted.

Reply:
Hmm, I don't know how that interacts with bluetooth devices. If someone figures it out later, they can adjust the comment